File: 8-string-to-integer-atoi/string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        i, n = 0, len(s)
        while i < n and s[i] == ' ':
            i += 1
        sign = 1
        if i < n and s[i] == '-':
            sign = -1
            i += 1
        elif i < n and s[i] == '+':
            i += 1
        result = 0
        while i < n and s[i].isdigit():
            digit = int(s[i])
            if result > (2 ** 31 - 1 - digit) // 10:
                return 2 ** 31 - 1 if sign == 1 else -2 ** 31
            result = result * 10 + digit
            i += 1
        return sign * result

File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0, n = s.size();
        while(i < n && s[i] == ' ') i++;
        int sign = 1;
        if(s[i] == '+' && i < n) i++;
        else if(s[i] == '-' && i < n){
            sign = -1;
            i++;
        }
        int num = 0;
        while(i < n && isdigit(s[i])){
            int digit = s[i] - '0';
            if(num > (INT_MAX - digit) / 10){
                return sign == 1 ? INT_MAX : INT_MIN;
            }
            num = 10 * num + digit;
            i++;
        }
        return num * sign;
    }
};